### PR TITLE
Weather

### PR DESCRIPTION
**Pull request by IlyPoh**

### CODE REVIEW

**Review by kosachevlad — COMMENTED** [review]
QA

1. First upload - svg to html in full screen
2. the preloader is already hidden, but there is no information on the screen yet
3. When switching to another city - a glitch - the layout goes somewhere
4. The first preloader should be displayed between requests (when switching cities in dropdown)
- switch to mobile resolution
5. when the menu is open, the weather card is not visible (2 options for solving: close on click outside, close on selecting)
6. the card is cropped when resolution 370px, the window cannot be scrolled
7. localStorage does not work as expected

[(Video) testing IlyPoh weather app
](https://youtu.be/xnaPgydeqi4)

**Review by kosachevlad — COMMENTED** [review]
i'd recommend you to change file structure of your project.
image

1. move all source files (html, scss, js) to src folder
2. in scripts folder create 2 folders (components, helpers)
3. also i'd recommend you to rename `scripts` folder to `js`, `script.js` file to `app.js`, `data.js` file to `config.js`

File: index.html [hunk 1]
```diff
@@ -4,11 +4,101 @@
...
+                <div class="weather-current-temperature">
+                    <div class="weather-image"></div>
+                    <div>
+                        <span class="weather-temperature"></span>
+                    </div>
+                </div>
+                <div class="weather-current-info"></div>
+                <div class="weather-current-additional-info">
+                    <div class="weather-current-section">
+                        <svg viewBox="0 0 1000 1000" enable-background="new 0 0 1000 1000" class="icon-direction-pointer">
```
Comment (kosachevlad) [inline index.html]: I'd recommend you to delete this svg's from source html, and import them from js using svg-inline-loader for example.

File: package.json [hunk 2]
```diff
@@ -0,0 +1,25 @@
+{
+    "name": "weather",
+    "version": "1.0.0",
+    "description": "",
+    "main": "index.js",
+    "scripts": {
+        "test": "echo \"Error: no test specified\" && exit 1",
```
Comment (kosachevlad) [inline package.json]: delete "test" script if you don't use it

File: scripts/city.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import { getLocalTime, dewPointCelsius, findWindDirection } from './helpers.js'
+export class CityInfo {
```
Comment (kosachevlad) [inline scripts/city.js]: use blank line here to format code
```suggestion

export class CityInfo {
```

File: scripts/city.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+        this.icon = data.weather[0].icon
+        this.time = getLocalTime(data.timezone)
+        this.feelsLike = Math.round(data.main.feels_like)
+        this.windDeg = data.wind.deg
+        this.windDirection = findWindDirection(data.wind.deg)
+        this.windSpeed = data.wind.speed
+        this.pressure = data.main.pressure
+        this.humidity = data.main.humidity
+        this.dew = dewPointCelsius(data.main.temp, data.main.humidity),
+        this.visibility = (data.visibility / visibilityMetric).toFixed(1)
```
Comment (kosachevlad) [inline scripts/city.js]: In my environment your application works only with this fix
```suggestion
        this.visibility = (data.visibility / CityInfo.visibilityMetric).toFixed(1)
```

File: scripts/helpers.js [hunk 5]
```diff
@@ -0,0 +1,64 @@
+import { months, directions } from "./data";
+
+export function createDOMElement(tagName, ...classNames) {
+    const element = document.createElement(tagName);
+    element.classList.add(...classNames);
+    return element;
```
Comment (kosachevlad) [inline scripts/helpers.js]: blank line before return
```suggestion

    return element;
```

File: scripts/helpers.js [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+    const date = localTime.getDate()
+    const month = localTime.getMonth()
+    const currentMonth = months[month]
+
+    const ampm = hours >= 12 ? 'PM' : 'AM'
+    hours = hours % 12
+    hours = hours ? hours : 12;
+
+    return `${hours}:${minutes} ${ampm} ${currentMonth} ${date}`
+}
```
Comment (kosachevlad) [inline scripts/helpers.js]: blank line after the function end
```suggestion
}

```

File: scripts/helpers.js [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+export function dewPointCelsius(temperature, humidity) {
+    const vaporPressureConstant = 17.27;
+    const saturationVaporPressureConstant = 237.7;
+    const alpha = ((vaporPressureConstant * temperature) / (saturationVaporPressureConstant + temperature)) + Math.log(humidity / 100.0);
+    const dewPoint = Math.round((saturationVaporPressureConstant * alpha) / (vaporPressureConstant - alpha));
+    return dewPoint;
```
Comment (kosachevlad) [inline scripts/helpers.js]: blank line before return
```suggestion

    return dewPoint;
```

File: scripts/helpers.js [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+export function findWindDirection(windDeg) {
+    const windDirectionDegrees = 45
```
Comment (kosachevlad) [inline scripts/helpers.js]: semicolon
```suggestion
    const windDirectionDegrees = 45;
```

File: scripts/helpers.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+    return windDirectionCardinal
+}
+
+export function gettingError(message) {
+    const errorMessageBlockElement = document.querySelector('.weather-error')
+    errorMessageBlockElement.classList.remove('hidden')
+    const errorMessageTextElement = document.createElement('p')
+    errorMessageTextElement.innerHTML = message
+    errorMessageBlockElement.append(errorMessageTextElement)
+    return;
```
Comment (kosachevlad) [inline scripts/helpers.js]: useless return

File: scripts/helpers.js [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+}
+
+export function firstCharToUpperCase(text) {
+    return text = text[0].toUpperCase() + text.slice(1)
+}
+
+export function responseErrorChecker(data) {
+    if (data.cod === 401) {
```
Comment (kosachevlad) [inline scripts/helpers.js]: typo in word code
```suggestion
    if (data.code === 401) {
```

File: scripts/helpers.js [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+        gettingError('Invalid API Key.')
+        return;
```
Comment (kosachevlad) [inline scripts/helpers.js]: useless return

File: scripts/index.js [hunk 12]
```diff
@@ -0,0 +1,2 @@
+import '../scss/app.scss'
+import {} from "./script"
```
Comment (kosachevlad) [inline scripts/index.js]: i don't think its a good practice to use {} as global import. You should define what function you want to export, or wrap it in one function (i.e. 'app') and export it as default.

File: scripts/script.js [hunk 13]
```diff
@@ -0,0 +1,127 @@
...
+function userGeolocation() {
+    navigator.geolocation.getCurrentPosition(
+        (data) => {
+            const lat = data.coords.latitude;
+            const lon = data.coords.longitude;
+            findCityByGeolocation(lat, lon)
+        },
+        (error) => {
+            gettingError(error.message)
+            if (!window.localStorage.getItem("userCity")) {
```
Comment (kosachevlad) [inline scripts/script.js]:
```suggestion
            if (!localStorage.getItem("userCity")) {
```

File: scripts/script.js [hunk 14]
```diff
@@ -0,0 +1,127 @@
...
+                findCityByName(cityList[0])
+            } else {
+                findCityByName(window.localStorage.getItem('userCity'))
```
Comment (kosachevlad) [inline scripts/script.js]:
```suggestion
                findCityByName(localStorage.getItem('userCity'))
```

File: scripts/script.js [hunk 15]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+function animation() {
+    const { loading } = elements
+    loading.classList.remove('is-loading')
+}
+
+function selectedCityUpdate() {
+    const { citySelected } = elements
+    citySelected.innerHTML = window.localStorage.getItem('userCity')
```
Comment (kosachevlad) [inline scripts/script.js]:
```suggestion
    citySelected.innerHTML = localStorage.getItem('userCity')
```

File: scripts/script.js [hunk 16]
```diff
@@ -0,0 +1,127 @@
...
+}
+
+async function findCityByName(cityName) {
+    const response = await fetch(`https://api.openweathermap.org/data/2.5/weather?q=${cityName}&appid=${apiKey}&units=metric`)
```
Comment (kosachevlad) [inline scripts/script.js]: I'd recommend you to store URL and strings in config.js file
```javascript
const URL = 'https://api.openweathermap.org/';
const fetchWeatherByCity = (city) => `${URL}data/2.5/weather?q=${city}&appid=${apiKey}&units=metric`;
const fetchWeatherByGeolocation = (lat, lon) => `${URL}data/2.5/weather?lat=${lat}&lon=${lon}&appid=${apiKey}&units=metric`;

export {URL, fetchWeatherByCity, fetchWeatherByGeolocation}
```
and import them in your current file
```javascript
import {URL, fetchWeatherByCity, fetchWeatherByGeolocation} from './components/config.js'
// ...
    const response = await fetch(`fetchWeatherByCity(cityName)`)
```

Comment (kosachevlad) [reply scripts/script.js]: Also you should wrap your async functions into try/catch block
```javascript
async function findCityByName(cityName) {
  try {
     const response = await fetch(`fetchWeatherByCity(cityName)`);
     const data = await response.json();
     if (!response.ok) {
         responseErrorChecker(data);
     }
     updater(data);
  } catch (error) {
      console.log(error);
  }
  ```

File: scripts/script.js [hunk 17]
```diff
@@ -0,0 +1,127 @@
...
+    const data = await response.json()
+    if (!response.ok) {
+        responseErrorChecker(data)
+    }
+    updater(data)
+}
+
+async function findCityByGeolocation(lat, lon) {
```
Comment (kosachevlad) [inline scripts/script.js]: same

File: scripts/script.js [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+
+function updater(data) {
+    const city = new CityInfo(data)
+
+    updateCity(city)
+    selectedCityUpdate()
+}
+
+function updateCity(city) {
+    elements.images.forEach( ele => {
```
Comment (kosachevlad) [inline scripts/script.js]: bad parameter naming. You already know that elements.images contains only images
```suggestion
    elements.images.forEach( img => {
```

File: scripts/script.js [hunk 19]
```diff
@@ -0,0 +1,127 @@
...
+        const imageElement = document.createElement('img')
+        imageElement.setAttribute('src', `https://openweathermap.org/img/wn/${city.icon}@2x.png`)
```
Comment (kosachevlad) [inline scripts/script.js]:
```suggestion
        imageElement.setAttribute('src', `${URL}img/wn/${city.icon}@2x.png`)
```

File: scripts/script.js [hunk 20]
```diff
@@ -0,0 +1,127 @@
...
+        ele.innerHTML = ""
+        ele.append(imageElement)
+    })
+    city.secondState = firstCharToUpperCase(city.secondState)
+    elements.direction.style.transform = `rotate(${city.windDeg}deg)`
+
+    elements.temperatures.forEach( ele => ele.innerHTML = `${city.temp}° C`)
```
Comment (kosachevlad) [inline scripts/script.js]: bad parameter naming. Make correct abbreviations of words
```suggestion
    elements.temperatures.forEach( elem => elem.innerHTML = `${city.temp}° C`)
```

File: scripts/script.js [hunk 21]
```diff
@@ -0,0 +1,127 @@
...
+
+document.addEventListener('click', (e) => {
+    if (e.target.classList.contains('weather-city-selected')) toggleList();
+    if (e.target.classList.contains('weather-city-city')) findCityByName(e.target.dataset.for);
+})
+
+document.addEventListener("DOMContentLoaded", async () => {
+    userGeolocation()
+    createCityList(cityList)
+    await new Promise(
```
Comment (kosachevlad) [inline scripts/script.js]: delete this inappropriate Promise. I already told you via telegram how to implement preloader behaviour.

**Review by kosachevlad — COMMENTED** [review]
also you should use prettier extension in vscode.

**Review by IlyPoh — COMMENTED** [review]
+animation fix

File: scripts/helpers.js (hunk 10 above)
Comment (IlyPoh) [reply scripts/helpers.js]: I got this response from the server.
image

**Review by kosachevlad — APPROVED** [review]

**Pull request merged**